### PR TITLE
persistentVector, persistentHashMap, persistentHashSet implementations

### PR DESCRIPTION
**Pull request by qurbonzoda**

### CODE REVIEW

**Comment by ilya-g** [comment]
Let's address those of existing tests that have been commented out in this PR.

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableMap/PersistentHashMapBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kotlinx.collections.immutable.implementations.immutableMap
+
+import kotlinx.collections.immutable.ImmutableMap
+
+class Marker
```
Comment (ilya-g) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableMap/PersistentHashMapBuilder.kt]: Use at most `internal` visibility.

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableMap/PersistentHashMapBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    override fun put(key: K, value: @UnsafeVariance V): V? {
+        val keyHash = key?.hashCode() ?: NULL_HASH_CODE
+        node = node.makeMutableFor(this)
+        return node.mutablePut(keyHash, key, value, 0, this)
+    }
+
+    override fun putAll(from: Map<out K, V>) {
+        from.forEach { key, value -> put(key, value) }
```
Comment (ilya-g) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableMap/PersistentHashMapBuilder.kt]: This won't compile with JDK 6, use `forEach` with a single parameter, which can be immediately destructured (or not):
```
from.forEach { (k, v) -> ... }
// or
from.forEach { put(it.key, it.value) }
```

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableMap/PersistentHashMapBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    override fun remove(key: K): V? {
+        val keyHash = key?.hashCode() ?: NULL_HASH_CODE
+        node = node.makeMutableFor(this)
```
Comment (ilya-g) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableMap/PersistentHashMapBuilder.kt]: Making node mutable may be premature if the key is not in the map.

File: kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/implementations/immutableMap/PersistentHashMapTest.kt [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+
+            map = map.remove(index)
+        }
+    }
+
+    @Test
+    fun collisionTests() {
+        var map = persistentHashMapOf<KeyWrapper<Int>, Int>()
+
+        assertEquals(2, map.put(KeyWrapper(1, 1), 1).put(KeyWrapper(1, 1), 2)[KeyWrapper(1, 1)])
```
Comment (ilya-g) [inline kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/implementations/immutableMap/PersistentHashMapTest.kt]: This isn't a collision, it's just a value update for a given key.

**Comment by ilya-g** [comment]
You can use `AbstractList/Set/Map` and `AbstractMutableList/Set/Map` from `kotlin.collections` as base types of your implementations freely.

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableMap/PersistentHashMapBuilder.kt (hunk 1 above)
Comment (qurbonzoda) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableMap/PersistentHashMapBuilder.kt]: +

File: kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/implementations/immutableMap/PersistentHashMapTest.kt (hunk 4 above)
Comment (qurbonzoda) [reply kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/implementations/immutableMap/PersistentHashMapTest.kt]: +1

**Comment by pakoito** [comment]
💯

**Pull request merged**